src/python/bot/fuzzers/mutator_plugin.py

```diff
@@ -0,0 +1,167 @@
+
+
+def _get_mutator_plugins_subdir(subdir):
+  """Returns the path of the subdirectory named |subdir| in
+  MUTATOR_PLUGINS_DIR."""
+  return os.path.join(environment.get_value('MUTATOR_PLUGINS_DIR'), subdir)
+
+
+def _get_mutator_plugins_archives_dir():
+  """Returns path of the archives subdirectory in MUTATOR_PLUGINS_DIR"""
+  return _get_mutator_plugins_subdir(ARCHIVES_SUBDIR_NAME)
+
+
+def _get_mutator_plugins_unpacked_dir():
+  """Returns path of the unpacked plugins subdirectory in MUTATOR_PLUGINS_DIR"""
+  return _get_mutator_plugins_subdir(PLUGINS_SUBDIR_NAME)
+
+
+def _get_mutator_plugins_from_bucket():
+  """Returns list of the mutator plugin archives in the mutator plugin storage
+  bucket."""
+  return storage.list_blobs(_get_mutator_plugins_bucket_url())
+
+
+def _download_mutator_plugin_archive(mutator_plugin_archive):
+  """Downloads the |mutator_plugin_archive| from the mutator plugin storage
+  bucket to the plugin archives directory. Returns the path that the archive was
+  downloaded to."""
+  file_path = os.path.join(_get_mutator_plugins_archives_dir(),
+                           mutator_plugin_archive)
+  url = '%s/%s' % (_get_mutator_plugins_bucket_url(), mutator_plugin_archive)
+  if not storage.copy_file_from(url, file_path):
+    logs.log_error(
+        'Failed to copy plugin archive from %s to %s' % (url, file_path))
+    return None
+
+  return file_path
+
+
+def _unpack_mutator_plugin(mutator_plugin_archive_path):
+  """Unpacks |mutator_plugin_archive_path| in the unpacked plugins directory and
+  returns the path it was unpacked into."""
+  mutator_plugin_name = os.path.basename(
+      os.path.splitext(mutator_plugin_archive_path)[0])
+  unpacked_plugin_dir = os.path.join(_get_mutator_plugins_unpacked_dir(),
+                                     mutator_plugin_name)
+  archive.unpack(mutator_plugin_archive_path, unpacked_plugin_dir)
+  return unpacked_plugin_dir
+
+
+def _extract_name_from_archive(plugin_archive_filename):
+  """Parses |plugin_archive_filename| which should be named using the schema:
+  '$NAME-$JOB-$FUZZ_TARGET.zip'. Returns tuple containing
+  ($NAME, $JOB-$FUZZ_TARGET)."""
+  # TODO(metzman): Get rid of this when we create an upload page for custom
+  # mutator plugins.
+  plugin_archive_name = os.path.splitext(plugin_archive_filename)[0]
+  idx = plugin_archive_name.index('-')
+  plugin_name = plugin_archive_name[:idx]
+  return plugin_name, plugin_archive_name[idx + 1:]
+
+
+class PluginGetter(object):
+  """Class that gets a usable mutator plugin for |fuzzer_binary_name| from
+  GCS."""
+
+  def __init__(self, fuzzer_binary_name):
+    self.fuzzer_binary_name = fuzzer_binary_name
+    self.job_name = environment.get_value('JOB_NAME')
+    self.create_directories()
+
+  def create_directories(self):
+    """Creates directories needed to use mutator plugins."""
+
+    def recreate_directory(directory_path):
+      shell.create_directory_if_needed(
```

Comment:
This looks very hacky, to do create_directory_if_needed first and then remove_directory. Why do we need to remove and re-download plugin everytime ? Is this for every launcher instance ? We need to make sure we dont even unpack if it was unpacked once.

Reply:
> This looks very hacky, to do create_directory_if_needed first and then remove_directory. Why do we need to remove and re-download plugin everytime ? 

What if the plugin gets updated?

>Is this for every launcher instance ? 

Yes.

>We need to make sure we dont even unpack if it was unpacked once.

I intentionally do this. I agree it would be less wasteful to unpack once for every 5 runs of launcher.py. I believe the code did this initially, but it was suggested that I change this so that mutator plugins would more easily work on OSS-Fuzz.